constraint/pkg/apis/constraints/apis.go

```diff
@@ -40,8 +66,102 @@ func GetEnforcementAction(constraint *unstructured.Unstructured) (string, error)
 	}
 
 	if !found {
-		return EnforcementActionDeny, nil
+		return string(Deny), nil
 	}
 
 	return action, nil
 }
+
+func IsEnforcementActionScoped(action string) bool {
+	return action == string(Scoped)
+}
+
+// GetEnforcementActionsForEP returns a map of enforcement actions for enforcement points passed in.
+func GetEnforcementActionsForEP(constraint *unstructured.Unstructured, eps []string) (map[string][]string, error) {
+	scopedActions, found, err := getNestedFieldAsArray(constraint.Object, "spec", "scopedEnforcementActions")
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", ErrInvalidSpecEnforcementAction, err)
+	}
+	if !found {
+		return nil, fmt.Errorf("%w: spec.scopedEnforcementActions must be defined", ErrMissingRequiredField)
+	}
+
+	scopedEnforcementActions, err := convertToSliceScopedEnforcementAction(scopedActions)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", ErrInvalidConstraint, err)
+	}
+
+	enforcementPointsToActionsMap := make(map[string]map[string]bool)
+	for _, ep := range eps {
+		enforcementPointsToActionsMap[ep] = make(map[string]bool)
+	}
+	for _, scopedEA := range scopedEnforcementActions {
+		for _, enforcementPoint := range scopedEA.EnforcementPoints {
+			epName := strings.ToLower(enforcementPoint.Name)
+			var action string
+			switch scopedEA.Action {
+			case string(Warn), string(Dryrun):
+				action = scopedEA.Action
+			default:
+				action = string(Deny)
```

Comment:
shouldnt we return error? we cant assume non valid values just all default to deny. we can only set to deny when enforcementAction is not provided.

Reply:
https://github.com/open-policy-agent/gatekeeper/blob/bd96c5263523be54642dcb4a88c2a9e502a74ea7/pkg/webhook/policy.go#L300-L309 this code assumes not dryrun, not warn, means deny because it's processing the response returned by the driver. but we should not make that assumption here. 

https://github.com/open-policy-agent/gatekeeper/blob/bd96c5263523be54642dcb4a88c2a9e502a74ea7/pkg/webhook/policy.go should also be updated to include scope otherwise, it will default to deny.

Reply:
Right now the behavior is, if someone turns off the `disable-enforcementaction-validation` flag on gatekeeper, then constraints can be created with whatever enforcementAction (`enforcementAction: fake`) - [Link](https://github.com/open-policy-agent/gatekeeper/blob/bd96c5263523be54642dcb4a88c2a9e502a74ea7/pkg/webhook/policy.go#L415). In this case, constraint gets added to the template client regardless of the "valid/invalid" enforcement action. Keeping the behavior same IMO we should just pass whatever scoped enforcement action is in the constraint without validating it when adding the constraint to template client. The mechanism to handle this already exists in gatekeeper when we receive responses in gatekeeper with "unrecognised" actions.

By default, we are validationg enforcement action and scoped enforcement actions to reject constraint with not valid enforcement action. `enforcementAction: deny, dryrun, warn, scoped` and `scopedEnforcementActions[].action: denym dryrun, warn`

 So should I remove this block altogether then? I dont think we need this check here.

Reply:
> https://github.com/open-policy-agent/gatekeeper/blob/bd96c5263523be54642dcb4a88c2a9e502a74ea7/pkg/webhook/policy.go should also be updated to include scope otherwise, it will default to deny.

I think we should keep this as is because when response include `scoped` enforcement action, the implementation looks at `scoped` enforcement actions. https://github.com/open-policy-agent/gatekeeper/pull/3321/files#diff-d3c8e52b699d3596d1057b82890a2ed8624445f1ee8f6d3d8a532e13766b99a8R261

Reply:
Rita is correct, the code should be removed.

The fact that constraints can be created with arbitrary enforcement actions is intentional. This is because we cannot know, and should not presume to know, what enforcement actions are available to any EPs.

We should not validate enforcement actions and just take them as we come.

If EPs want to perform additional validation (e.g. Gatekeeper raising a warning/reporting via `status` if it is on the hook for taking an action it does not recognize), that's fine, but CF is not equipped to do that. At best it can provide helper functions to allow EPs to implement that extra porcelain.

Assuming unrecognized actions == deny is not a good assumption to make. Example: if Gatekeeper adds a PubSub action, the CF would rewrite it as deny.

Reply:
Also, we should not return an error, as I alluded to earlier, CF does not and should not know the full set of valid enforcement actions, it just forwards them to EPs, who are responsible for interpreting them.